csrc/id_graphs.h

```diff
@@ -0,0 +1,465 @@
+#pragma once
+
+#include <disjoint_set.h>
+#include <ir_all_nodes.h>
+#include <kernel_ir.h>
+#include <lower_trivial_broadcast.h>
+
+#include <deque>
+#include <unordered_map>
+
+namespace nvfuser {
+
+using IdGroup = std::shared_ptr<VectorOfUniqueEntries<IterDomain*>>;
+using IdGroups = VectorOfUniqueEntries<IdGroup>;
+using ExprGroup = std::shared_ptr<VectorOfUniqueEntries<Expr*>>;
+using ExprGroups = VectorOfUniqueEntries<ExprGroup>;
+
+class TORCH_CUDA_CU_API IdGraph {
+ public:
+  IdGraph() = default;
+
+  IdGraph(const IdGraph& other);
+  IdGraph(IdGraph&& other) = default;
+
+  IdGraph& operator=(const IdGraph& other);
+  IdGraph& operator=(IdGraph&& other) = default;
+
+  // Returns the disjoint IterDomain set.
+  const DisjointSets<IterDomain*>& disjointIdSets() const;
+
+  DisjointSets<IterDomain*>& disjointIdSets();
+
+  // Returns
+  //   {
+  //     (1) The disjoint set of the provided Iter Domain if it exists,
+  //     otherwise a null shared ptr
+  //     (2) If the disjoint set of the provided Iter Domain exists
```

Comment:
Returning an optional probably makes a lot of sense, but actually I think most instances I just want to assert there's actually an id set. So probably I'll just put an alias.

Reply:
I just stole the pattern for https://cplusplus.com/reference/unordered_set/unordered_set/emplace/ for development, no reason not to switch to optional.

Reply:
Ah OK. Makes sense. I think `std::optional` was introduced in C++17, but `std::unordered_set` has been around since C++11 so they had to use the explicit `pair<..., bool>` . From side discussion it looks like we might be OK with using C++17 going forward on NVFuser. Either way not a big deal of course.